### DISCUSSION TITLE
[API Request] Control / stream scrobling action

### DISCUSSION THREAD

**Discussion by Tolriq** [opensubsonic/open-subsonic-api#22] [API requests]
### Type of change

API Clarification

### Proposal description

IMO this is an API clarification but it can also be a API change depending on the feedback of current servers.

While this is not officially documented some servers currently handle an access to `stream` as a playcount. 

This is wrong because it can be just used to play 1 sec of content, or used to download an offline media.

Since this is not documented this can also trigger duplicate playcount increase when clients properly send a scrobble after.

TL;DR: OpenSubsonic servers should not increase playcount on `stream` access and rely on the `scrobble` api that is meant for that.

### Backward compatibility impact

Depends on current servers and client assumption.

### Backward compatibility

- [ ] No backward compatibility impact.

### API details

N/A

### Security impacts

N/A

### Alternative solutions

The alternative solution is to add a new parameter to `stream` so that clients can specifically opt out of that mode and control manually via `scrobble` endpoint

**Comment by gravelld** [comment]
This is interesting - it almost needs the server to interrogate the user agent to know when to increment the play count. Otherwise you'll never get any play counts if the client doesn't implement the newer API.

**Reply by Tolriq** [reply]
If your server rely on that, then the alternative solution is to add an opt-out parameter to the `stream` endpoint. That was my initial proposal in Symfonium forum, but @deluan mentioned that it actually did not really make sense to count playcount by just an access to the `stream` endpoint.

**Reply by dweymouth** [reply]
I should hope that all clients send the scrobble request when they want to register a play count increase. Any client that doesn't do this is wrong and I think it's ok if that client doesn't increase play counts after OpenSubsonic is rolled out. That way its users will file a bug report and the client dev will know to add the scrobble.

**Reply by Tolriq** [reply]
@gravelld Reading again your message ` if the client doesn't implement the newer API.` the `scrobble` api is not new at all. And the goal of the OpenSubsonic is to never have to  check the server name or the clients header to take decisions. Every details should be documented to avoid current mess and improve things.

**Reply by gravelld** [reply]
Sorry - you're right. I completely forgot about the existing scrobble API.

**Comment by lachlan-00** [comment]
I 100% vote for this. The stream endpoint is annoying as hell especially as a dsub user.

I have an option in ampache to ignore/allow scrobble on stream so it could be retooled to enable for older clients.

**Comment by Tolriq** [comment]
Ok so there was quite a few request this is nice, now we need to figure out the voting system and organization to start building on those requests.

As a start since this is the most upped request and the simplest one since everyone seems OK with it, let's try to define the voting and validation process.

@opensubsonic/servers @opensubsonic/clients :)

For the vote I propose to use reactions as it's easy to count and see who vote.
 👍 🚀 would be a positive vote.

👎 ⛔ would be a negative vote.

For the rest it's up for debate. I suppose we should have 2 "rules" 
1) There's only positive votes then TBD: 
- How many server and client do we need to validate?
2) There's at least one negative vote: Should server negative vote have more weight than client vote? How do we deal with ? 
- Tweak until no negative vote?
- Consider X +1 vote erase 1 negative vote?
- Something else?

**Comment by ghenry22** [comment]
This is more a clarification than anything else the original subsonic and all direct forks recorded a play on request to the stream end point.

The documentation for the scrobble endpoint is not clear and the behaviour of the stream endpoint is not noted in the API doc though so confusion isn’t a surprise.

Newer servers which were not just direct forks (ie navidrome) choose to use the scrobble end point to record plays.

I remember I had to update substreamer for this use case.

Assuming we start at a new API version (say 2.0) for the first round of changes it’s our chance to create some clearer updated docs and note new endpoints or changes each version.  We could actually use proper semantic versioning which would be nice…

It would be great to get some of the non-breaking / smaller updates into a first version so we can get it implemented across all our platforms and a first release out, ironing out the process and showing users we are on top of things :)

**Reply by Tolriq** [reply]
Hum you should read the few posts about how it was voted to build OpenSubsonic. There's no 2.0 it's based on extensions, there's already doc online about how it will work.

This particular endpoint is easy and self contained, so this is more about validating the voting process than anything.

**Reply by ghenry22** [reply]
Sure I guess I missed that original discussion and to be honest I have no strong opinion on it, so happy to go whichever way.

just made sense to me to continue to increment API level as it makes it easy to enable or disable features and the mechanism already exists.  If the base API level is the last level supported by the official subsonic then we can build where ever we want after that and still maintain base compatibility.

edit: to be clear I don’t mean rebuild the whole API from scratch I just meant increase the API level and put an agreed set of new features or changes in each API level increase.  Not too many at once so everyone can implement and stick together.

**Comment by sentriz** [comment]
i have seen a lot of clients that have a "scrobble to lastfm (if supported by server)" checkbox. which is unchecked by default. does this mean that we wont ever get playcount data if they don't check the box? imagine people without a lastfm account wouldn't think to check it

**Reply by Tolriq** [reply]
I don't know how the other clients works so can't really tell, but from the current votes and upvotes the non playcount on stream access is largely prefered.

See for example https://github.com/kagemomiji/airsonic-advanced/issues/275

Any access triggers playcount, downloading a file when transcoding triggers a playcount, seeking in the file triggers a playcount if it restart a connection. This makes the playcount value completely absurd and not usable in many cases.
Including offline plays that are not counted.

It's better to have clients fix their side if they rely on this behavior that only some servers do, and have a single well documented usage of scrobble that does it's job.

**Reply by ghenry22** [reply]
Play count on stream API call was the behaviour in the original subsonic and is the behaviour in direct forks (madsonic, airsonic).

As subsonic is the “source” technically that is then the correct behaviour and servers that implement otherwise are non-compliant with the subsonic API.

However, I agree that a sensible clarification and standardisation in the way playcount and the scrobble api work and the removal of auto playcount increment in stream is a good way to go.

Navidrome already has a good implementation of this.

any client that current records plays with navidrome should just work, otherwise it’s a pretty trivial update to call the scrobble api when appropriate.

**Reply by Tolriq** [reply]
Yes and it's a pre requisite for https://github.com/opensubsonic/open-subsonic-api/discussions/21 that is also a big pain and issue for tons of users.

Needs approval of the PR .

**Reply by ghenry22** [reply]
Added my approval from client group.

**Comment by lachlan-00** [comment]
To make this happen I think we need to extend stream or mandate use of download.

In ampache I have a cache parameter (e.g. cache=1) because dsub doesn't use the download method

Download isn't a stream so clients should be using that to avoid a stream playcount.

**Reply by Tolriq** [reply]
This is actually already voted. Subsonic calls on stream should not count by default. There's the scrobble endpoint for that.

Download endpoint does not work for any transcoded downloads either, stream endpoint calls for seeking or reloading or many other cases triggers wrong playcounts too.
